runtime-sdk/src/error.rs

```diff
@@ -29,26 +29,171 @@ use crate::types::transaction::CallResult;
 /// ```
 pub trait Error: std::error::Error {
     /// Name of the module that emitted the error.
-    fn module_name() -> &'static str;
+    fn module_name(&self) -> &'static str;
 
     /// Error code uniquely identifying the error.
     fn code(&self) -> u32;
 
     /// Converts the error into a call result.
     fn to_call_result(&self) -> CallResult {
-        CallResult::Failed {
-            module: Self::module_name().to_owned(),
-            code: self.code(),
+        // If there is an underlying error, use that one.
+        match Error::source(self) {
+            Some(source) => source.to_call_result(),
+            None => CallResult::Failed {
+                module: self.module_name().to_owned(),
+                code: self.code(),
+            },
         }
     }
+
+    /// The lower-level source of this error, if any.
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        None
+    }
 }
 
 impl Error for std::convert::Infallible {
-    fn module_name() -> &'static str {
+    fn module_name(&self) -> &'static str {
         "(none)"
     }
 
     fn code(&self) -> u32 {
         Default::default()
     }
 }
+
+#[doc(hidden)]
+pub mod private {
+    use super::Error;
+
+    pub trait AsDynError {
+        fn as_dyn_error(&self) -> &(dyn Error + 'static);
+    }
+
+    impl<T: Error + 'static> AsDynError for T {
+        #[inline]
```

Comment:
I would hope that everyone uses LTO. The kind of person who doesn't use LTO on release builds is the same one who declares `struct None`.